src/mode/terminal.rs

```diff
@@ -302,104 +302,206 @@ where
+            [0x1f, 0x20, 0x10, 0x10, 0x20, 0x1f],
+            // X
+            [0x21, 0x12, 0x0c, 0x0c, 0x12, 0x21],
+            // Y
+            [0x01, 0x02, 0x3c, 0x02, 0x01, 0x00],
+            // Z
+            [0x21, 0x31, 0x29, 0x25, 0x23, 0x21],
+            // [
+            [0x3f, 0x21, 0x00, 0x00, 0x00, 0x00],
+            // \
+            [0x02, 0x04, 0x08, 0x10, 0x20, 0x00],
+            // ]
+            [0x21, 0x3f, 0x00, 0x00, 0x00, 0x00],
+            // ^
+            [0x04, 0x02, 0x3f, 0x02, 0x04, 0x00],
+            // _
+            [0x40, 0x40, 0x40, 0x40, 0x40, 0x40],
+            // `
+            [0x01, 0x02, 0x00, 0x00, 0x00, 0x00],
+            // a
+            [0x10, 0x2a, 0x2a, 0x2a, 0x3c, 0x00],
+            // b
+            [0x3f, 0x24, 0x24, 0x24, 0x18, 0x00],
+            // c
+            [0x1c, 0x22, 0x22, 0x22, 0x00, 0x00],
+            // d
+            [0x18, 0x24, 0x24, 0x24, 0x3f, 0x00],
+            // e
+            [0x1c, 0x2a, 0x2a, 0x2a, 0x24, 0x00],
+            // f
+            [0x00, 0x3e, 0x05, 0x01, 0x00, 0x00],
+            // g
+            [0x18, 0xa4, 0xa4, 0xa4, 0x7c, 0x00],
+            // h
+            [0x3f, 0x04, 0x04, 0x04, 0x38, 0x00],
+            // i
+            [0x00, 0x24, 0x3d, 0x20, 0x00, 0x00],
+            // j
+            [0x20, 0x40, 0x40, 0x3d, 0x00, 0x00],
+            // k
+            [0x3f, 0x0c, 0x12, 0x20, 0x00, 0x00],
+            // l
+            [0x1f, 0x20, 0x20, 0x00, 0x00, 0x00],
+            // m
+            [0x3e, 0x02, 0x3c, 0x02, 0x3c, 0x00],
+            // n
+            [0x3e, 0x02, 0x02, 0x02, 0x3c, 0x00],
+            // o
+            [0x1c, 0x22, 0x22, 0x22, 0x1c, 0x00],
+            // p
+            [0xfc, 0x24, 0x24, 0x24, 0x18, 0x00],
+            // q
+            [0x18, 0x24, 0x24, 0x24, 0xfc, 0x00],
+            // r
+            [0x3e, 0x04, 0x02, 0x02, 0x00, 0x00],
+            // s
+            [0x24, 0x2a, 0x2a, 0x2a, 0x10, 0x00],
+            // t
+            [0x02, 0x1f, 0x22, 0x20, 0x00, 0x00],
+            // u
+            [0x1e, 0x20, 0x20, 0x20, 0x1e, 0x00],
+            // v
+            [0x06, 0x18, 0x20, 0x18, 0x06, 0x00],
+            // w
+            [0x1e, 0x30, 0x1c, 0x30, 0x1e, 0x00],
+            // x
+            [0x22, 0x14, 0x08, 0x14, 0x22, 0x00],
+            // y
+            [0x1c, 0xa0, 0xa0, 0xa0, 0x7c, 0x00],
+            // z
+            [0x22, 0x32, 0x2a, 0x26, 0x22, 0x00],
+            // {
+            [0x0c, 0x3f, 0x21, 0x00, 0x00, 0x00],
+            // |
+            [0x3f, 0x00, 0x00, 0x00, 0x00, 0x00],
+            // }
+            [0x21, 0x3f, 0x0c, 0x00, 0x00, 0x00],
+            // ~
+            [0x02, 0x01, 0x02, 0x01, 0x00, 0x00],
+            // blank
+            [0x00, 0x00, 0x00, 0x00, 0x00, 0x00],
+        ];
+
+        let g = if ('!'..='~').contains(&input) {
+            CHARS[(input as u8 - b'!') as usize]
+        } else {
+            CHARS[CHARS.len() - 1]
+        };
```

Comment:
This could be written a bit more elegantly like this:

```suggestion
        let g = CHARS
            .get((input as u8 - b'!') as usize)
            .unwrap_or(&CHARS[CHARS.len() - 1]);
```

It also saves a few bytes in my testing but YMMV! Can you confirm a size reduction in your setup?

Reply:
Comitted this, but reverted because if input is smaller than '!' (for example space or newline) it leads to a panic: https://app.circleci.com/pipelines/github/jamwaffles/ssd1306/620/workflows/40c25c3e-6806-4af1-a7e7-679410eb9579/jobs/5370

Reply:
Right, sorry about that - I wasn't able to run the tests when I made this suggestion. How about this:

```suggestion
        let g = (input as usize)
            .checked_sub(b'!'.into())
            .and_then(|idx| CHARS.get(idx))
            .unwrap_or(&CHARS[CHARS.len() - 1]);
```

The `checked_sub` returns `None` on underflow, so won't panic any more. I ran a few tests and this seems to work pretty nicely. It also still helps the binary size compared to the original code which is good.

Reply:
Committed. Thanks, TIL!

I've checked the x86 assembly output for my version vs your version and your version is definitely more compact. Is it because this way we express intent more clearly and rust compiler is capable to trigger some extra optimisation techniques?

Reply:
> Is it because this way we express intent more clearly and rust compiler is capable to trigger some extra optimisation techniques?

Yeah, something like that :D. I don't have a very deep understanding of how all this works at the low level but a rule of thumb I follow is if I can chain some methods instead of doing it imperitively (`if` statements, etc), Rust will _likely_ be able to do more optimisations with them and create faster code as it can make more assumptions, or can discard invariants it knows can't exist any more, etc.